Fix Add & Edit TT to have target value change HBT instead of sensitivity

mountrcg opened the pull request:
To fix issue #829 and replace #716.

needs extensive testing. Basically it only calculates the sensitivity (Insulin percentage) when the target value is choosen and the current HBT is the same as the one in Settings, which is always the case when you switch from a standard sensitivity TT to a customized one. After that, when changing target values the current state sensitivity is kept and the HBT calculated to be aligned for this new TT and current sensitivity.

mountrcg committed:
fix TT to have target value only change HBT

mountrcg commented:
min HBT input in settings guard rails are set to 105. Currently 100 is allowed, which is also not correct and gives inplausible results if chosen.

mountrcg committed:
change lowest HBT input to 105

mountrcg commented:
We need another general computation guardrail for low HBT in settings, which can leed to sensitivity <15%. Eg HBT set to 105 and then a TT is set at 195. Resulting sensRatio would be 5%, which is currently still set for standard and custom highTT's.
In those cases the used hbt needs to deviate from setting HBT in order to raise sensRatio to 15%. Seem complicated especially for a standard highTT that raises sens in this case.

kingst commented:
@mountrcg looks like this is still a draft, let me know when you're ready for a review and I can take a look.

mountrcg commented:
Honestly, I haven’t figured out how to change the HBT for a standard temp target, which would be a required for a settings HBT of 105 and the target value above 135. @marv-out  did these parts. Normaly a standard TT does not get a a custom HBT, but in order not to get below 15% some edge cases would requiere it - messy. May be we accept that as is?

dnzxy commented:
@mountrcg , the logic you are talking about can be found here https://github.com/nightscout/Trio/blob/dev/Trio/Sources/APS/OpenAPS/OpenAPS.swift#L536-L552

I'm not 100% certain that I understand the edge case you have found with a setting-HBT of 105, but this is what I understood:

Given a normal target and a half-basal target, we would normally compute the adjustment sensitivity and based on this new sensitivity, manipulate the setting-HBT with a custom-HBT. But if the computed sensitivity would be < 15%, then the sensitivity must be capped at 15%, and in that capped case, we want to "recompute" the HBT value for that same target, such that the resulting sensitivity is exactly 15%.

If that's correct, we could do something like the following:
1. Add a new helper function.
2. Recompute the HBT.
3. Check if the recomputed HBT differs from the temp target HBT, if so, adjust HBT _again_.
4. Send the twice-adjusted HBT to oref, where the actual sensitivity adjustment takes place.

Add this to `OpenAPS.swift`
```swift
private func computeCappedHalfBasalTarget(
    normalTarget: Decimal,
    tempTarget: Decimal,
    originalHBT: Decimal
) -> Double {
    // Step 1: compute P from existing HBT
    // P = (HBT - Normal) / (HBT - Temp + Normal)
    let numerator = originalHBT - normalTarget
    let denominator = originalHBT - tempTarget + normalTarget
    
    guard denominator != 0 else {
        return round(Double(originalHBT))  // just be safe
    }

    let computedRatio = numerator / denominator
    let computedPercent = Double(truncating: computedRatio as NSNumber) * 100
    
    // Step 2: check for <15%
    if computedPercent >= 15 {
        return round(Double(originalHBT))
    }

    // Step 3: cap at 15% and recompute HBT
    let cappedRatio = Decimal(0.15)

    let cappedHBT =
        ((2 * cappedRatio * normalTarget) - normalTarget - (cappedRatio * tempTarget)) /
        (cappedRatio - 1)

    return round(Double(cappedHBT))
}
```

The computation logic was taken from `AdjustmentsStateModel%2BExtensions/AdjustmentsStateModel%2BTempTargets.swift`, reversed and adds capping.

Below can be found in `OpenAPS.swift`.
```swift
// Check if a Temp Target is active and if its HBT differs from user preferences
if let activeTempTarget = try self.fetchActiveTempTargets().first,
   activeTempTarget.enabled,
   let activeHBT = activeTempTarget.halfBasalTarget?.decimalValue,
   activeHBT != defaultHalfBasalTarget
{
   const recomputedHBT = self.computeCappedHalfBasalTarget(
            normalTarget: normalTarget, // grab via profile
            tempTarget: activeTempTarget.target,
            halfBasalTarget: activeHBT
        )

    // Overwrite the HBT in preferences
    adjustedPreferences.halfBasalExerciseTarget = activeHBT != recomputedHBT ? recomputedHBT : activeHBT
    debug(.openAPS, "Updated halfBasalExerciseTarget to active Temp Target value: \(activeHBT)")
}
```

This is rough code, just written down, while looking at the few files that are involved here. Not yet tested, just roughly outlined.

mountrcg commented:
> I'm not 100% certain that I understand the edge case you have found with a setting-HBT of 105, but this is what I understood:
> 
> Given a normal target and a half-basal target, we would normally compute the adjustment sensitivity and based on this new sensitivity, manipulate the setting-HBT with a custom-HBT. But if the computed sensitivity would be < 15%, then the sensitivity must be capped at 15%, and in that capped case, we want to "recompute" the HBT value for that same target, such that the resulting sensitivity is exactly 15%.

Not quite. We have 2 kinds of TT with adjusted sensitivity.
1. standard
2. customized

For (2) we do not calculate adjustedSensitivity in this PR, as per @kingst suggestion we always take that value as input and calculate the HBT that will realize this. So problem solved as the input is guarded.

For (1) we take the target value and the HBT from settings and calculate adjusted sensitivity. So here is the issue
a. There is no check against the 15%. 
b. If there would be, we then need to calculate a new HBT to use instead of the one from settings. Also fine and now in PR, but as far as I know the standard TT does not carry over a custom HBT to oref, only if TT is "Custom".

So when I test the the standard TT with that low HBT all looks fine in the Add/Eddit TT but oref will calculate the 11% sens adjustment.

<img width="400" alt="Simulator Screenshot - Dev-16Pro - 2025-12-08 at 18 13 20" src="https://github.com/user-attachments/assets/a7594445-fd57-40ff-8f25-67fc255802bc" />
<img width="400" alt="Simulator Screenshot - Dev-16Pro - 2025-12-08 at 18 13 02" src="https://github.com/user-attachments/assets/9b3627a4-fc3b-4383-8b9f-0b3bc6cdd92b" />

dnzxy commented:
@mountrcg I‘m aware how it works. My outlining addressed both cases. Have a read through the code.

mountrcg commented:
here are the 2 functions that calc

1. adjustedSens https://github.com/nightscout/Trio/blob/3328da60515af9f3c06b5599ecacfb19fec27980/Trio/Sources/Modules/Adjustments/AdjustmentsStateModel%2BExtensions/AdjustmentsStateModel%2BTempTargets.swift#L429
2 adjustedHBT https://github.com/nightscout/Trio/blob/3328da60515af9f3c06b5599ecacfb19fec27980/Trio/Sources/Modules/Adjustments/AdjustmentsStateModel%2BExtensions/AdjustmentsStateModel%2BTempTargets.swift#L388

So I move those to helpers and use in OpenAPS?

dnzxy commented:
My suggestion was to add a new computation function as a helper util for OpenAPS. You could probably use those two functions and add some logic around them in OpenAPS as a different way to solving this, either or :)

mountrcg committed:
check in OpenAPS when enacting presets

move HBP and Sens calculations to Helper file

mountrcg committed:
use helper constants

mountrcg committed:
make sure guarded percentage is shown in HomeView

mountrcg commented:
I have reduced the PR and only added the guarding of min Insulin% of 15% in OpenAPS and the adjusted Insulin% in HomeView. Also moved the HBT and Sens calculations to a helper file to use in:

- Add / Edit TT Forms and the States
- OpenAPS
- HomeView

Ready for testing and Review @dnzxy & @kingst

mountrcg commented:
Test cases:
Try to get Insulin% substantially below 15% by 

1. changing the target values of new or edited high TT's (custom or standard)
2. setting the HBT in settings to it's minimum of 105 and adding new high TT's > 140, or testing existing TT's > 140

dnzxy commented on Trio/Sources/APS/OpenAPS/OpenAPS.swift:
```diff
@@ -534,15 +534,26 @@ final class OpenAPS {
         // Check for active Temp Targets and adjust HBT if necessary
         try await context.perform {
-            // Check if a Temp Target is active and if its HBT differs from user preferences
+            // Check if a Temp Target is active
```
We do still check if the HBT differs (just differently now) and we technically still adjust it here, if need be.

mountrcg replied:
done

dnzxy commented on Trio/Sources/APS/OpenAPS/OpenAPS.swift:
```diff
@@ -534,15 +534,26 @@ final class OpenAPS {
-                debug(.openAPS, "Updated halfBasalExerciseTarget to active Temp Target value: \(activeHBT)")
+                // Compute effective HBT - handles both custom HBT and standard TT (where HBT might need adjustment)
+                let effectiveHBT = TempTargetCalculations.computeEffectiveHBT(
+                    storedHBT: activeTempTarget.halfBasalTarget?.decimalValue,
```
Please rename `storedHalfBasalTarget` to `tempTargetHalfBasalTarget` for all instances + comments, because _stored_ could be misread as the actual setting-stored HBT.

mountrcg replied:
done

dnzxy commented on Trio/Sources/Helpers/TempTargetCalculations.swift:
```diff
@@ -0,0 +1,132 @@
+    /// The normal target glucose value used as reference (100 mg/dL)
+    static let normalTarget: Decimal = 100
+
+    /// Computes the raw (unclamped) adjusted percentage for a given HBT and target.
```
Do we gain anything by having these two functions present separately and not just always having the low-end clamp present? This guard / checking condition could be part of `computeRawAdjustedPercentage`, could it not?

dnzxy commented on Trio/Sources/Helpers/TempTargetCalculations.swift:
```diff
@@ -0,0 +1,132 @@
+    ///   - target: The target glucose value
+    ///   - autosensMax: The maximum autosens multiplier from settings
+    /// - Returns: A tuple containing (percentage, halfBasalTarget) where halfBasalTarget is nil if settings HBT is OK
+    static func computeStandardPercentageAndHBT(
```
If we combine the two functions, this function here because much easier &/or even because a no longer needed wrapper-util-function, because instead of doing what this function here encapsulates, you might as well just add an invocation of those 2 function that are being called within this one.

dnzxy commented on Trio/Sources/Helpers/TempTargetCalculations.swift:
```diff
@@ -0,0 +1,132 @@
+    ///   - target: The target glucose value
+    ///   - autosensMax: The maximum autosens multiplier from settings
+    /// - Returns: The effective HBT to use, or nil if settings HBT should be used as-is
+    static func computeEffectiveHBT(
```
Again, a wrapper, for a wrapper, that just encapsulates a lot of stuff.
Do we need 1-2 layers of abstraction for something as simple as getting a % value and a target value?

mountrcg replied:
done

dnzxy commented on Trio/Sources/Models/DecimalPickerSettings.swift:
```diff
@@ -72,7 +72,7 @@ struct DecimalPickerSettings {
         value: 160,
         step: 5,
-        min: 100,
+        min: 105,
```
Let's please add a short inline-comment here referencing/explaining, that 105 is a hard-coded oref limit and shall not be adjusted (because it will/can lead to inconsistent behavior when calculating TTs).

mountrcg replied:
done

dnzxy commented on Trio/Sources/Modules/Adjustments/AdjustmentsStateModel+Extensions/AdjustmentsStateModel+TempTargets.swift:
```diff
@@ -421,24 +418,74 @@ extension Adjustments.StateModel {
 
+    /// Computes the raw (unclamped) adjusted percentage for a given HBT and target.
+    /// This is used to check if the percentage would be below minSensitivityRatioTT.
+    func computeRawAdjustedPercentage(
```
Why are you abstracting this by this wrapper instead of just directly calling `TempTargetCalculations.computeRawAdjustedPercentage(…)`? This seems unnecessary.

mountrcg replied:
done

dnzxy commented on Trio/Sources/Modules/Adjustments/AdjustmentsStateModel+Extensions/AdjustmentsStateModel+TempTargets.swift:
```diff
@@ -421,24 +418,74 @@ extension Adjustments.StateModel {
+    }
+
     /// Computes the adjusted percentage for the slider.
     func computeAdjustedPercentage(
```
Same question here? Why abstract this instead of just directly calling TempTargetCalculations util?

mountrcg replied:
done

dnzxy commented on Trio/Sources/Modules/Adjustments/AdjustmentsStateModel+Extensions/AdjustmentsStateModel+TempTargets.swift:
```diff
@@ -421,24 +418,74 @@ extension Adjustments.StateModel {
+    /// returns an adjusted HBT that yields the minimum percentage instead.
+    /// - Parameter target: The target glucose value
+    /// - Returns: A tuple containing (percentage, halfBasalTarget) where halfBasalTarget is nil if standard HBT can be used
+    func computeStandardPercentageAndHBT(usingTarget target: Decimal) -> (percentage: Double, halfBasalTarget: Decimal?) {
```
Same question here? Either, just call the function in `TempTargetCalculations` that does _exactly this_ and returns a tuple of percentage and HBT value, or (if you are removing that layer of abstraction, because I asked for it in my comments in `TempTargetCalculations`), why not call the 2 util functions here directly? Why abstract this instead of just directly calling it? Do we gain anything by having a tuple instead of two "unconnected" values?

dnzxy reviewed (commented):
Reviewed it and ran it briefly. It seems okay functionality wise, but I did have a lot of questions. The coding choices done here seem to add a lot of complexity via introducing abstraction layers / encapsulation through wrapper functions, where your initial aim (as per @kingst 's comments on #716) was to massively simplify all this logic. 

Please don't understand this as criticism, just trying to get to the easiest result, hence my questions :) 

Thanks for tackling this and looking into fixing this edge case.

PS: Maybe we should a add a semi-long + descriptive docstring _somewhere_ explaining why 
- 15% is a hard-coded minimum limit for sensitivity adjustments
- why ASmin and ASmax are not used the same way when clamping TT sensitivity adjustments
- where this stems from (i.e., oref / AndroidAPS)

mountrcg committed:
rename to tempTargetHalfBasalTarget

mountrcg committed:
correct comment

mountrcg committed:
Simplify by making raw percentage calculation private

TempTarget sensitivity calculations into TempTargetCalculations helper

mountrcg committed:
remove wrapper in State model and call directly

mountrcg committed:
refactor: computeEffectiveHBT

remove computeStandardPercentageAndHBT wrapper, inline logic into computeEffectiveHBT

mountrcg committed:
comment on scoping/boundary of HBT

mountrcg committed:
remove remaining state calls in UI files

call TempTargetCalculations directly

mountrcg committed:
add comment for HBT boundaries and additional explaination for TT's

mountrcg committed:
standardize debug message

mountrcg commented:
tackled all the valueble comments and reduced the logging now

mountrcg committed:
add more values to debug

dnzxy commented on Trio/Sources/Modules/Adjustments/AdjustmentsStateModel+Extensions/AdjustmentsStateModel+TempTargets.swift:
```diff
@@ -138,6 +138,11 @@ extension Adjustments.StateModel {
         let date = self.date
         guard date > Date() else { return }
 
+        let adjustmentType = halfBasalTarget == settingHalfBasalTarget ? "Standard" : "Custom"
```
Does a (static) variable for this helper variable make sense vs. adding the same snippet 3 times?

mountrcg replied:
I would like to leave it, used it often searching in code

dnzxy commented on Trio/Sources/Modules/Adjustments/View/TempTargets/TempTargetHelpView.swift:
```diff
@@ -7,20 +7,45 @@ struct TempTargetHelpView: View {
+                .listRowBackground(Color.gray.opacity(0.1))
+
+                Section {
+                    VStack(alignment: .leading, spacing: 10) {
```
Following text suggestions (via GPT, adjusted for high school student literacy).
Feel free to use/adjust.

```swift
VStack(alignment: .leading, spacing: 10) {
    Text(
        "Sensitivity changes from Temp Targets have a built-in minimum of 15%. Even very high Temp Targets cannot reduce insulin delivery below 15% of normal."
    )
    Text(
        "This 15% minimum is a safety limit taken from oref (OpenAPS reference design) and AndroidAPS. It helps prevent insulin delivery from dropping to unsafe levels."
    )
    Text(
        "Important: Autosens Min and Autosens Max do not affect Temp Targets in the same way. Autosens Max limits how much insulin can be increased, but Autosens Min does not remove the 15% minimum when insulin is reduced."
    )
    Text(
        "This difference exists because situations like exercise often need a much larger insulin reduction than Autosens would detect during a normal daily routine."
    )
}
```

dnzxy reviewed (approved):
Two small comments, but overall I'm okay with this.
@kingst please give this a review. I think we're at a good point and have simplified things.

kingst reviewed (approved):
LGTM

mountrcg committed:
adjust help text to suggestion by Dan
